docs/architecture/adr-001-autopilot-oci-basic-auth-support.md

```diff
@@ -0,0 +1,137 @@
+        URL string `json:"url"`
+
+        // Sha256 provides an optional SHA256 hash of the URL's content for verification.
+        Sha256 string `json:"sha256,omitempty"`
+
+        // ArtifactPullSecrets holds a reference to a secret where the credentials are
+        // stored. We use these credentials when pulling the artifacts from the provided
+        // URL using any of the supported protocols (http, https, and oci).
+        ArtifactPullSecret *ArtifactPullSecret `json:"artifactPullSecret,omitempty"`
+
+        // Insecure indicates whether certificates in the remote URL (if using TLS) can
+        // be ignored, aka InsecureSkipVerify.
+        Insecure bool  `json:"insecure,omitempty"`
+}
+```
+
+The `ArtifactPullSecret` property will be added and its struct will be defined as follow:
+
+```go
+type ArtifactPullSecret struct {
+      // Namespace of the secret.
+      Namespace string `json:"namespace"`
+
+      // Name of the secret.
+      Name string `json:"name"`
+}
+```
+
+The secret pointed by the provided `ArtifactPullSecret` will be used for pulling artifacts using either HTTP[S] or OCI protocols and is expected to by of type `kubernetes.io/dockerconfigjson` if the protocol in use is `oci://` or of type `Opaque` if protocols `http://` or `https://` are used (see below for details on the Secret layout).
+
+Example configuration for OCI:
+
+```yaml
+url: oci://my.registry/binaries/k0s:v1.30.1+k0s.0
+sha256: e95603f167cce6e3cffef5594ef06785b3c1c00d3e27d8e4fc33824fe6c38a99
+artifactPullSecret:
+  namespace: kube-system
+  name: artifacts-registry
+```
+
+Example configuration for HTTPS:
+
+```yaml
+url: https://my.file.server/binaries/k0s-v1.30.1+k0s.0
+sha256: e95603f167cce6e3cffef5594ef06785b3c1c00d3e27d8e4fc33824fe6c38a99
+artifactPullSecret:
+  namespace: kube-system
+  name: artifacts-basic-auth
+```
+
+Example configuration for HTTP:
+
+```yaml
+url: http://my.file.server/binaries/k0s-v1.30.1+k0s.0
+sha256: e95603f167cce6e3cffef5594ef06785b3c1c00d3e27d8e4fc33824fe6c38a99
+artifactPullSecret:
+  namespace: kube-system
+  name: artifacts-token-based-auth
+```
+
+### Secrets Layout
+
+For secrets of type `kubernetes.io/dockerconfigjson` the format is the default for Docker authentications, equal to what is used in a Pod's pull secret. For further details you can refer to the [official documentation](https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/).
+
+When it comes to the `Opaque` secret layout (used for HTTP requests) Autopilot will accept the following entries:
+
+- `username` and `password`: if both are set then Autopilot will attempt to pull the artifacts using [Basic Authentication](https://www.ibm.com/docs/en/cics-ts/6.1?topic=concepts-http-basic-authentication).
+- `authorization`: if this property is set then the `Authorization` [header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Authorization) will be set to its value when pulling the artifacts.
```

Comment:
Is there any possibility, that different registries might use different headers, e.g. `X-Auth-Token`, `X-Access-Token`, etc.? Or accept some additional headers? 

Maybe we can have a secret type, that will put all the secret content directly into headers without any processing.

Reply:
In my opinion we don't need to worry about this for registry access. Registries, more or less, work in the same way and we have implementation of registry clients for go. The `authorization` property here will be used only for HTTP or HTTPS requests.